Improve memory reuse efficiency and reduce page faults if using two-level hash tables

jiebinn opened the pull request:
This patch will change the default `lg_extent_max_active_fit` from 6 to 8. It will enhance hot dirty memory reuse when using two-level hashtables. 

**Performance issue analysis:**
We identified the performance issue in many of the 43 ClickBench queries with ClickHouse on the 2x240 vCPUs system, particularly high page faults. For a deeper investigation, let's consider Q35. Q35 exhibited around 40% `__handle_mm_fault` hotspot in cycles on the GNR 2x240 vCPUs platform. We discovered that the high page faults stem from `MADV_DONTNEED` with the `perf event`.
In Query 35, there are 256 memory reallocations (sub-hashtables) from 4KB to 16KB for each arena. Jemalloc recycles and coalesces many of these 256 16K memory blocks into a larger one when ClickHouse frees them with the `bpftrace` data. Subsequently, the large memory space in the dirty ecache cannot be reused for the next following 16K request, as the maximum allowed memory space before splitting is 16 * 64 KB. When a new memory request is made, Jemalloc locates an existing extent larger than the requested size in the dirty ecache and splits it to fit the actual requested size, maximizing memory reuse. There is a boundary for the requested extent size and the existing extent size, with a maximum ratio of 64 to minimize memory fragmentation. Consequently, Jemalloc would find possible memory space from the retained ecache (already `MADV_DONTNEED` and lacking physical pages) or through mmap, resulting in page faults and high RSS.

**What does the jemalloc patch do:**
The 256 pieces of 16K size dirty memory will coalesce into a memory block larger than 64*16K, preventing reuse when a new 16K request arrives. To maximize dirty ecache reuse, we can increase the maximum ratio of the existing extent size to the requested extent size from 64 to 256.

**Ref:** 
https://github.com/jemalloc/jemalloc/pull/2842

**Result:**
We have tested this patch with Clickbench Q35 on a system with 2 x 240 vCPUs. The results show significant performance gains (opt/base):

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/jiebinsu/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/jiebinsu/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{border:.5pt solid windowtext;}
.xl66
	{color:black;
	vertical-align:middle;
	border:.5pt solid windowtext;}
-->
</style>
</head>

<body link="#0563C1" vlink="#954F72">


Q35 | Query per second | VmRSS | Total page fault   (50 runs) | cycles | instructions | IPC
-- | -- | -- | -- | -- | -- | --
Opt/Base | 1.961 | 0.546 | 0.29 | 0.43 | 0.857 | 1.993



</body>

</html>


The geometric mean of all 43 queries shows more than a 10% performance improvement.

Refs: jemalloc/jemalloc#2842

<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Performance Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve memory reuse efficiency and reduce page faults when using the twolevel hashtables.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

rschu1ze commented:
Does this PR update to a development version of jemalloc?

jiebinn commented:
> Does this PR update to a development version of jemalloc?

This PR will update to a development version of jemalloc. There is a significant performance improvement when queries use two-level hash tables for aggregation. Each thread will have 256 sub-hash tables for each two-level hash table. For example, if the initial size of a sub-hash table is 4K (256 cells * 16 bytes/cell) and it needs to reallocate from 4K to 16K (1024 cells * 16 bytes/cell) based on historical cache data, there will be 256 reallocation requests (one for each sub-hash table) from 4K to 16K sent to jemalloc. This can cause 3X page faults, 2X VmRSS, 50% QPS.

To avoid these issues, if we don't update the jemalloc version, we might need to work around by passing variables in settings to jemalloc or changing the management of memory blocks in two-level hash tables. However, both approaches may introduce other concerns.

By the way, do we have any tests if we decide to update the submodule version? I also noticed that the latest jemalloc version ([jemalloc @ 41a859e]) in ClickHouse is on the dev branch, which is fine as jemalloc is a mature and stable allocator.

rschu1ze commented:
@jiebinn First of all, thanks for the PR.

> This PR will update to a dev version of jemalloc.

That's unfortunately a problem for ClickHouse, especially for something so fundamental as memory management. The [jemalloc](https://jemalloc.net/) homepage says (not surprisingly) "dev: The dev branch tracks current development, and at any given time may not be production-ready.". Unfortunately, they seem to make stable [releases](https://github.com/jemalloc/jemalloc/releases) only rarely.

However, this is also only my opinion, I can't tell how reliable dev versions of jemalloc are really.
Let me tag @antonio2368 for a second opinion (he is on a longer leave right now, so might post in 1-2 months only).

> By the way, do we have any tests if we decide to update the submodule version? I also noticed that the latest jemalloc version ([jemalloc @ 41a859e]) in ClickHouse is on the dev branch, which is fine as jemalloc is a mature and stable allocator.

This is the exact version which ClickHouse is using:

```
41a859ef 2022-07-02 14:44:46 -0400 Jasmin Parent (HEAD) Remove duplicated words in documentation
```

It is only ten (or so) commits ahead of tag `5.3.0` (from 2022-05-06 11:28:25). I am actually not sure if this was a coincidence or if it means "dev versions are okay and we merely don't track dev HEAD closely".

rschu1ze commented:
@jiebinn Yes, but only if it has no dependencies to previous commits on the development branch and if it is straightforward. We should avoid using ClickHouse as a canary for finding bugs in jemalloc.

Algunenano commented:
I think using dev branches it's ok, but it depends on the project policy. If their dev branch is considered stable and final (think of absl for example) then there is no issue. If their dev branch is a candidate, which they will cut at some point and then fix bugs before release, then it's probably not ok. In general (but it's my experience) jemalloc is usually rock solid, so I expect it to be on the first case.

OTOH, running any branch in CI and compare / detect issues is ok. Based on findings and results we can decide whether the risk is worth or not.

jiebinn commented:
@rschu1ze , I agree with that if we cherry-pick one commit temporarily, the patch should be clear and simple, and in good quality.

@Algunenano , I believe jemalloc is quite solid. I agree with that we can decide whether to update to a dev branch based on our findings and CI or other test results.

I'll first ask jemalloc if there are any plans to release a new stable version soon, as that would be the best solution. If there are no plans, we might consider cherry-picking this commit (https://github.com/jemalloc/jemalloc/pull/2842), using the dev branch, or implementing a workaround in ClickHouse.

rschu1ze commented:
Okay, tnx.

@jiebinn To let our CI test your PR, please fix the build issues ... once all builds are green, functional/perf/end-to-end tests will start.

But by the looks of things, it seems the [build problems](https://s3.amazonaws.com/clickhouse-test-reports/PRs/80245/25594832709d8fbe6de369396ba96a4934aed392//build_amd_binary/job.log) are in jemalloc itself. To fix, feel free to replace the official jemalloc submodule (`.gitmodules` in the ClickHouse repository) by a forked one which fixes the build.

Also, to fix the build, please open a build log (e.g. the one I linked above ^^), then grep for `cmake `. This will find something like this:

```
Run command: [   cmake --debug-trycompile -DCMAKE_VERBOSE_MAKEFILE=1 -LA -DCMAKE_BUILD_TYPE=None  -DENABLE_THINLTO=0 -DSANITIZE=          -DENABLE_CHECK_HEAVY_BUILDS=1 -DENABLE_CLICKHOUSE_SELF_EXTRACTING=1 -DCMAKE_C_COMPILER=clang-19 -DCMAKE_CXX_COMPILER=clang++-19 -DCOMPILER_CACHE=sccache -DENABLE_BUILD_PROFILING=1 -DENABLE_TESTS=0 -DENABLE_UTILS=0 -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON /home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse]
```

which you can run locally to reproduce (only remove trash like `DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY` and `DENABLE_CLICKHOUSE_SELF_EXTRACTING`)

jiebinn commented:
Hi @rschu1ze and @Algunenano ,we can have a safer and more convenient method to address the high page faults issue when queries use two-level hash tables in ClickHouse. Previously, the default `lg_extent_max_active_fit` was set to 6, resulting in a maximum ratio of 64 between the size of a dirty merged memory extent and the size of the requested allocation. Since two-level hash tables in ClickHouse send 256 reallocation requests, the coalesced large memory block cannot be reused as it will usually exceed the max ratio of 64. To better fit our actual scenario in ClickHouse, we can increase this ratio to 256.

Algunenano commented:
@jiebinn it seems you removed the update of the submodule in one of the last force pushes, so the PR is not  doing anything except changing `JEMALLOC_CONFIG_MALLOC_CONF`

jiebinn commented:
> @jiebinn it seems you removed the update of the submodule in one of the last force pushes, so the PR is not doing anything except changing `JEMALLOC_CONFIG_MALLOC_CONF`

Yes. This PR will change the `JEMALLOC_CONFIG_MALLOC_CONF`. It offers a low risk and convenient method to revolve the high page faults issue resulting from the two-level hash tables.

jiebinn commented on contrib/jemalloc-cmake/CMakeLists.txt:
```diff
@@ -34,7 +34,7 @@ if (OS_LINUX)
     # MADV_DONTNEED. See
     # https://github.com/ClickHouse/ClickHouse/issues/11121 for motivation.
-    set (JEMALLOC_CONFIG_MALLOC_CONF "percpu_arena:percpu,oversize_threshold:0,muzzy_decay_ms:0,dirty_decay_ms:5000,prof:true,prof_active:false,background_thread:true")
+    set (JEMALLOC_CONFIG_MALLOC_CONF "percpu_arena:percpu,oversize_threshold:0,muzzy_decay_ms:0,dirty_decay_ms:5000,prof:true,prof_active:false,background_thread:true,lg_extent_max_active_fit:8")
```
@Algunenano Maybe we should consider applying the change to both Linux and Non-Linux system. I will check the CI result.

jiebinn committed:
Improve memory reuse efficiency and reduce page faults

This patch helps to set lg_extent_max_active_fit to 8, which
will help jemalloc to reuse existing dirty extents more
efficiently when using two-level hash tables (256 sub hashtables
and reallocations). We have tested this patch with Clickbench Q35
on a system with 2 x 240 vCPUs. The results show significant
performance gains (opt/base):

- QPS: 1.96x
- VmRSS: 54.6%
- Page faults: 29%
- Cycles: 43%
- Instructions: 85.7%
- IPC: 1.99x

The geometric mean of all 43 queries shows more than a 10%
performance improvement.

Refs: jemalloc/jemalloc#2842

Signed-off-by: Jiebin Sun <jiebin.sun@intel.com>
Reviewed-by: Tianyou Li <tianyou.li@intel.com>
Reviewed-by: Wangyang Guo <wangyang.guo@intel.com>
Reviewed-by: Zhiguo Zhou <zhiguo.zhou@intel.com>

jiebinn committed:
Merge branch 'ClickHouse:master' into updateJemalloc

Algunenano commented:
@jiebinn I'm trying to reproduce the improvements in the description but I don't see pretty much any change:

> Result:
> We have tested this patch with Clickbench Q35 on a system with 2 x 240 vCPUs. The results show significant performance gains (opt/base):
> <style> </style>
> Q35	Query per second	VmRSS	Total page fault (50 runs)	cycles	instructions	IPC
> Opt/Base	1.961	0.546	0.29	0.43	0.857	1.993
> The geometric mean of all 43 queries shows more than a 10% performance improvement.

Testing with a `AMD Ryzen 9 7950X3D 16-Core Processor` (with hyperthreading enabled, so 32 vCPU): 

```
$ clickhouse benchmark --port 49000 --query "SELECT ClientIP, ClientIP - 1, ClientIP - 2, ClientIP - 3, COUNT(*) AS c FROM hits GROUP BY ClientIP, ClientIP - 1, ClientIP - 2, ClientIP - 3 ORDER BY c DESC LIMIT 10;" -i 50
```

* Master: localhost:49000, queries: 50, QPS: 4.925, RPS: 492470445.132, MiB/s: 1878.626, result RPS: 49.248, result MiB/s: 0.002.
* Changes: localhost:49000, queries: 50, QPS: 4.951, RPS: 495046115.319, MiB/s: 1888.451, result RPS: 49.506, result MiB/s: 0.002.


```
$ clickhouse benchmark --port 49000 --query "SELECT ClientIP, ClientIP - 1, ClientIP - 2, ClientIP - 3, COUNT(*) AS c FROM hits GROUP BY ClientIP, ClientIP - 1, ClientIP - 2, ClientIP - 3 ORDER BY c DESC LIMIT 10 settings max_threads=64;" -i 50
```

Master: localhost:49000, queries: 50, QPS: 1.695, RPS: 169451896.802, MiB/s: 646.408, result RPS: 16.946, result MiB/s: 0.001.
Changes: localhost:49000, queries: 50, QPS: 1.554, RPS: 155365551.705, MiB/s: 592.673, result RPS: 15.537, result MiB/s: 0.001.


I'm waiting for the perf report to run again, but if you could explain how/what are you measuring that'd be great. Does this problem only appear when running with large amount of threads (`2x240 vCPUs`)?


I see this in the build:
```
$ ack JEMALLOC_CONFIG_MALLOC_CONF
CMakeCache.txt
816:JEMALLOC_CONFIG_MALLOC_CONF_OVERRIDE:STRING=

contrib/jemalloc-cmake/include_linux_x86_64/jemalloc/internal/jemalloc_internal_defs.h
409:#define JEMALLOC_CONFIG_MALLOC_CONF "percpu_arena:percpu,oversize_threshold:0,muzzy_decay_ms:0,dirty_decay_ms:5000,prof:true,prof_active:false,background_thread:true,lg_extent_max_active_fit:8"
```

So I'm assuming it's applied correctly.

jiebinn committed:
Merge branch 'ClickHouse:master' into updateJemalloc

jiebinn commented:
> @jiebinn I'm trying to reproduce the improvements in the description but I don't see pretty much any change:
> 
> > Result:
> > We have tested this patch with Clickbench Q35 on a system with 2 x 240 vCPUs. The results show significant performance gains (opt/base):
> > <style> </style>
> > Q35	Query per second	VmRSS	Total page fault (50 runs)	cycles	instructions	IPC
> > Opt/Base	1.961	0.546	0.29	0.43	0.857	1.993
> > The geometric mean of all 43 queries shows more than a 10% performance improvement.
> 
> Testing with a `AMD Ryzen 9 7950X3D 16-Core Processor` (with hyperthreading enabled, so 32 vCPU):
> 
> ```
> $ clickhouse benchmark --port 49000 --query "SELECT ClientIP, ClientIP - 1, ClientIP - 2, ClientIP - 3, COUNT(*) AS c FROM hits GROUP BY ClientIP, ClientIP - 1, ClientIP - 2, ClientIP - 3 ORDER BY c DESC LIMIT 10;" -i 50
> ```
> 
> * Master: localhost:49000, queries: 50, QPS: 4.925, RPS: 492470445.132, MiB/s: 1878.626, result RPS: 49.248, result MiB/s: 0.002.
> * Changes: localhost:49000, queries: 50, QPS: 4.951, RPS: 495046115.319, MiB/s: 1888.451, result RPS: 49.506, result MiB/s: 0.002.
> 
> ```
> $ clickhouse benchmark --port 49000 --query "SELECT ClientIP, ClientIP - 1, ClientIP - 2, ClientIP - 3, COUNT(*) AS c FROM hits GROUP BY ClientIP, ClientIP - 1, ClientIP - 2, ClientIP - 3 ORDER BY c DESC LIMIT 10 settings max_threads=64;" -i 50
> ```
> 
> Master: localhost:49000, queries: 50, QPS: 1.695, RPS: 169451896.802, MiB/s: 646.408, result RPS: 16.946, result MiB/s: 0.001. Changes: localhost:49000, queries: 50, QPS: 1.554, RPS: 155365551.705, MiB/s: 592.673, result RPS: 15.537, result MiB/s: 0.001.
> 
> I'm waiting for the perf report to run again, but if you could explain how/what are you measuring that'd be great. Does this problem only appear when running with large amount of threads (`2x240 vCPUs`)?
> 
> I see this in the build:
> 
> ```
> $ ack JEMALLOC_CONFIG_MALLOC_CONF
> CMakeCache.txt
> 816:JEMALLOC_CONFIG_MALLOC_CONF_OVERRIDE:STRING=
> 
> contrib/jemalloc-cmake/include_linux_x86_64/jemalloc/internal/jemalloc_internal_defs.h
> 409:#define JEMALLOC_CONFIG_MALLOC_CONF "percpu_arena:percpu,oversize_threshold:0,muzzy_decay_ms:0,dirty_decay_ms:5000,prof:true,prof_active:false,background_thread:true,lg_extent_max_active_fit:8"
> ```
> 
> So I'm assuming it's applied correctly.

Hi @Algunenano, the PR is not related to the system's core number. If each sub-hashtable in the two-level hashtables falls within the size range of jemalloc-defined large extents (16KB to SC_LARGE_MAXCLASS), the PR would improve performance, especially when the sub-hashtable size exceeds 16KB but isn't excessively large. The closer the size is to 16KB, the greater the performance improvement, up to 2X QPS performance improvement at 16KB. If the allocation size exceeds the upper limit, jemalloc will use a different huge extent code path.
If Q35 with `hits` is running on a 64 vCPU system, the allocation size of sub-hashtables in the two-level hashtables is about 7X to 8X larger than on a 480 vCPU system. If the size exceeds the jemalloc-defined large extent size, or if the size is much larger than the lower limit of 16KB, the performance improvement is not significant. I wonder if more filters would help reduce the table size during the aggregation stage, which might help you reproduce the performance result.
We have also written a micro benchmark to reproduce the same scenario.

jiebinn committed:
Merge branch 'ClickHouse:master' into updateJemalloc

Algunenano commented:
* [AST fuzzer (amd_ubsan)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=80245&sha=d7378af5572a8d950f58b8c04d7d15e8016b7907&name_0=PR&name_1=AST%20fuzzer%20%28amd_ubsan%29) -> https://github.com/ClickHouse/ClickHouse/issues/80875
* [AST fuzzer (amd_debug)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=80245&sha=d7378af5572a8d950f58b8c04d7d15e8016b7907&name_0=PR&name_1=AST+fuzzer+%28amd_debug%29) -> https://github.com/ClickHouse/ClickHouse/issues/79451
* CH Inc sync -> Known issue
* [Stress test (amd_tsan)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=80245&sha=d7378af5572a8d950f58b8c04d7d15e8016b7907&name_0=PR&name_1=Stress%20test%20%28amd_tsan%29) -> https://github.com/ClickHouse/ClickHouse/issues/75961

No noticeable changes in performance tests; I only see it in the microbenchmarks. Still it seems ok to include and analyze in larger perf tests as part of the release

azat commented:
@jiebinn while it seems it works for 240x2 CPUs with one query it may increase memory fragmentation for more common setups, w/o giving any benefit, so I propose reverting it - https://github.com/ClickHouse/ClickHouse/pull/116836
